### ISSUE TITLE
Keras loss does not match loss computed from predictions

### ISSUE THREAD

**Issue by raspstephan** [keras-team/keras#12580]
I have noticed a difference between the loss that Keras returns during training and model.evaluate and what I compute myself from the predictions. 
```
import numpy as np
import keras
from keras.layers import Dense
from keras.models import Sequential
import keras.backend as K

X = np.random.rand(10000)
Y = X + np.random.rand(10000) / 5

X_train, X_valid = X[:8000], X[8000:]
Y_train, Y_valid = Y[:8000], Y[8000:]

model = Sequential([
    Dense(1, input_shape=(1,), activation='linear'),
])
model.compile('adam', 'mae')
model.fit(X_train, Y_train, epochs=1, batch_size=1024, validation_data=(X_valid, Y_valid))
>> Train on 8000 samples, validate on 2000 samples
>> Epoch 1/1
>> 8000/8000 [==============================] - 0s 51us/step - loss: 0.1900 - val_loss: 0.1845

model.evaluate(X_valid, Y_valid)
>> 0.18448934221267702

preds = model.predict(X_valid)
np.abs(Y_valid - preds).mean()
>> 0.4463157837393756
```
I have noticed the same for different architectures, data and loss functions (e.g. MSE). How does Keras compute the loss and how can I reproduce it from the predictions?

Keras version : 2.2.4
Tensorflow version: 1.12.0

**Comment by jvishnuvardhan** [comment]
@raspstephan I am getting different results when compared to you but the result from prediction and evaluation are almost same (upto 8 decimals).

model.evaluate(X_valid, Y_valid)
>> 1.1385564470291138

preds = model.predict(X_valid)
np.abs(Y_valid - preds).mean()
>> 1.1385564454936303
Keras and tf versions are same as yours 2.2.4, 1.12.0.

I have ran with keras (2.2.4), TF(1.13.1). Similar to the above, results were matching upto 8 decimals.

Could you reset the runtime and run the code again and see whether there is any mismatch?
If you still have the mismatch, can you run it in Google colab?

Let me know how it progresses. Thanks!

**Comment by SCP-173-cool** [comment]
I got the right results with your code.
![image](https://user-images.githubusercontent.com/22118890/55270555-11b68300-52db-11e9-9c73-76c926fa35ea.png)

**Comment by raspstephan** [comment]
I updated tensorflow to 1.13.1 and the problem went away. Thanks for looking into it.

**Comment by benbuleong** [comment]
I have a similar issue, except that I am using a custom loss and metrics. I upgraded to tensorflow 1.13.1 as well as keras 2.2.4, but it is still weird: model.evaluate and model.predict using the same metrics and dataset should give the same results, but the difference is too great to be correct.
```
def f1(y_true, y_pred):
    y_pred = K.round(y_pred)
    tp = K.sum(K.cast(y_true*y_pred, 'float'), axis=0)
    tn = K.sum(K.cast((1-y_true)*(1-y_pred), 'float'), axis=0)
    fp = K.sum(K.cast((1-y_true)*y_pred, 'float'), axis=0)
    fn = K.sum(K.cast(y_true*(1-y_pred), 'float'), axis=0)

    p = tp / (tp + fp + K.epsilon())
    r = tp / (tp + fn + K.epsilon())

    f1 = 2*p*r / (p+r+K.epsilon())
    f1 = tf.where(tf.is_nan(f1), tf.zeros_like(f1), f1)
    return K.mean(f1)

def f1_loss(y_true, y_pred):
    
    tp = K.sum(K.cast(y_true*y_pred, 'float'), axis=0)
    tn = K.sum(K.cast((1-y_true)*(1-y_pred), 'float'), axis=0)
    fp = K.sum(K.cast((1-y_true)*y_pred, 'float'), axis=0)
    fn = K.sum(K.cast(y_true*(1-y_pred), 'float'), axis=0)

    p = tp / (tp + fp + K.epsilon())
    r = tp / (tp + fn + K.epsilon())

    f1 = 2*p*r / (p+r+K.epsilon())
    f1 = tf.where(tf.is_nan(f1), tf.zeros_like(f1), f1)
    return 1 - K.mean(f1)

callbacks = [EarlyStopping(monitor='val_f1', patience=15),
             ModelCheckpoint(filepath=dname+'_best_model.h5', monitor='val_f1', save_best_only=True, mode='max', verbose=1)]

opt = Adam(lr=0.0001, beta_1=0.9, beta_2=0.999, decay=0.01)
model.compile(loss=f1_loss, optimizer=opt, metrics=[f1])
model.fit([X_train_audio, X_train_video, X_train_text], Y_train, batch_size=128, epochs=30, callbacks=callbacks, validation_data=([X_test_audio, X_test_video, X_test_text],Y_test))

model.evaluate([X_test_audio, X_test_video, X_test_text],Y_test, batch_size=128, verbose=1)

Out: [0.6195135793534878, 0.5118402310086592]

predictions = model.predict([X_test_audio, X_test_video, X_test_text])
sess = tf.InteractiveSession()
print (sess.run(f1(Y_test,predictions)))

Out: 0.17375563
```

**Comment by PGuti** [comment]
Possible explanation? .evaluate() will do an average of the metric across batches of batch_size, while predict will run predictions then your loss is computed over the whole array. This is quite a change in your case because average(f1(batch)) is quite different from f1(all_data).

What is the total test size ?
